### ISSUE TITLE
False negative `trailing-comma-tuple`

### ISSUE THREAD

**Issue by mattclay** [PyCQA/pylint#8201]
### Bug description

A false negative with the `trailing-comma-tuple` rule can occur due to code preceding the use of a comma.

```python
"""detected.py"""
# pylint: disable=expression-not-assigned

print("demo"),
```

```python
"""undetected.py"""
# pylint: disable=expression-not-assigned

DEMO = True

print("demo"),
```

### Command used

```shell
pylint detected.py undetected.py
```

### Pylint output

```shell
************* Module detected
detected.py:4:0: R1707: Disallow trailing comma tuple (trailing-comma-tuple)
```

### Expected behavior

************* Module detected
detected.py:4:0: R1707: Disallow trailing comma tuple (trailing-comma-tuple)
************* Module detected
undetected.py:6:0: R1707: Disallow trailing comma tuple (trailing-comma-tuple)

### Pylint version

```shell
pylint 2.16.1
astroid 2.14.1
Python 3.11.1 (main, Jan 28 2023, 18:47:50) [GCC 11.3.0]
```

### OS / Environment

Ubuntu 22.04.1

**Comment by Pierre-Sassoulas** [comment]
Wow, this one has WTF written all over it. (Thank you for creating the issue.)

**Comment by nickdrozd** [comment]
Adding a comma after the variable leads to the warning getting raised:

```python
# pylint: disable=expression-not-assigned

DEMO = True,

print("demo"),
```

Apparently the warning is raised from a function that processes tokens:

```python
    def process_tokens(self, tokens: list[tokenize.TokenInfo]) -> None:
        # Process tokens and look for 'if' or 'elif'
        for index, token in enumerate(tokens):
            token_string = token[1]
            if token_string == "elif":
                # AST exists by the time process_tokens is called, so
                # it's safe to assume tokens[index+1] exists.
                # tokens[index+1][2] is the elif's position as
                # reported by CPython and PyPy,
                # token[2] is the actual position and also is
                # reported by IronPython.
                self._elifs.extend([token[2], tokens[index + 1][2]])
            elif _is_trailing_comma(tokens, index):
                if self.linter.is_message_enabled("trailing-comma-tuple"):
                    self.add_message("trailing-comma-tuple", line=token.start[0])
```

I think what's happening is that the token processor doesn't really understand the structure of the code, it's just looking for where the commas are. Tuples can be broken up across multiple lines, so it has to look back to previous lines, looking for certain keywords.

Maybe the token-processing approach has been pushed a little too far.

**Comment by nickdrozd** [comment]
Just to finish the thought, one of the keywords the token processor currently looks for is the character `=`. That is `=` _at the textual level_. The `pylint: disable=` pragma contains that character, so its particular placement can affect whether the warning is emitted. Thus this one raises no warnings:

```python
print("demo"),  # pylint: disable = expression-not-assigned
```

While this one gets the trailing comma warning:

```python
# pylint: disable = expression-not-assigned
print("demo"),
```

No warning:

```python
# pylint: disable = expression-not-assigned

DEMO = True

print("demo"),
```

Gets the warning:

```python
# pylint: disable = expression-not-assigned

DEMO = True

#=
print("demo"),
```

**Comment by mattclay** [comment]
@nickdrozd Ah, that explains something else we were seeing. This isn't reported:

```python
"""demo.py"""
# pylint: disable=expression-not-assigned

try:
    pass
except KeyboardInterrupt:
    print("not reported"),
```

But this is:

```python
"""demo.py"""
# pylint: disable=expression-not-assigned

try:
    pass
except KeyboardInterrupt:
    print("is = reported"),
```

**Comment by Pierre-Sassoulas** [comment]
It sounds like we do not need a mere bug fix but a proper rework using the ast instead ?

**Comment by nickdrozd** [comment]
Thinking about it further, I don't think it would be possible to implement this check with the AST. The presence or absence of particular characters isn't available, because the AST is A. It may be crummy, but I think this will have to be done at the token level. (I could be wrong, of course. How do Flake et al handle this?)

@mattclay In the meantime, you might consider enabling `expression-not-assigned`. Technically, the expression `print('demo'),` evaluates to a tuple of one element (namely `None`). As far as Pylint is concerned, it's just a tuple. But the tuple expression isn't assigned to anything, which is suspicious. So if that warning were enabled, there would at least be some kind of indication that something is amiss.
